### DISCUSSION TITLE
This repo already use the Atenttion.py optimized?

### DISCUSSION THREAD

**Discussion by ZeroCool22** [hlky/stable-diffusion-webui#795] [Q&A] [answered]
https://github.com/Doggettx/stable-diffusion/blob/main/ldm/modules/attention.py

https://github.com/Doggettx/stable-diffusion/blob/main/ldm/modules/diffusionmodules/model.py

If not,  i can just replace this files and good to go?

Thx.

**Comment by EliEron** [comment, marked as answer]
It does not currently use the updated attention.py, though there are PRs open to integrate it.

Replacing the files should work fine. It's what I do currently, and I have not run into any problems.

**Reply by thedyze** [reply]
Been replacing these files also. While higher resolutions and quite nice performance is possible, I've been noticing a definite reduction in quality of the outputs. 
In fact the higher the resolution, the poorer quality of the output. Would investigate results thoroughly before integrating this.
It should be said that I'm only using his attention.py and model.py, not the other files in his commits.

**Reply by mchaker** [reply]
What do you mean by quality of the output?

Blurring? Random objects? Melted appearance?

**Reply by thedyze** [reply]
yup, melted appearance

**Reply by thedyze** [reply]
An example [img2img]
![Untitled-1](https://user-images.githubusercontent.com/72066289/189547006-e17a4d8c-6a98-4349-be13-4a645255416a.png)

**Reply by thedyze** [reply]
Ok Doggettx gave me a hint that higher guidance was needed at higher resolution. 
Been experimenting with 40-50 now, and am getting better results. 
I've also been using low res low guidance txt2img images as input in img2img. It seems this does not work that well at higher resolutions, a more detailed input image creates a lot better results. 

So may not be a problem with his files after all.

**Reply by kwaad2** [reply]
See, the issue with what you are speaking of, is that you are comparing different resolutions, and the entire rendering "vision" changes with the resolution, you might as well be changing the seed between those img2img.


Rendering a 512x512 on "stock" SD, I got `1.97it/s`
Rendering a 512x512 on Doggettx SD, I got `2.05it/s`  
That's nearly 5% faster.
(Pug1, is "stock", and Pug2 is Doggettx)
As far as I can tell they visually they are pixel perfect. 
![pug1](https://user-images.githubusercontent.com/28718864/189552710-25769442-1cf0-4fe7-9223-0b787473c16d.png)
![pug2](https://user-images.githubusercontent.com/28718864/189552712-1c890cb1-92a7-44a1-80fb-89292af44697.png)

Image Pug3, we see "stock" SD rendering a pug at 448x448, and it looks NOTHING like the pug rendered at 512x512. I am unable to render any higher than 512x512 with my RTX2060, so I had to go lower to "show' that  changing the resolution effectively changes the seed also.

![pug3](https://user-images.githubusercontent.com/28718864/189553110-30d48783-1c24-4b61-b8db-ddc2496e33c6.png)

Image Pug4, se seed Doggettx SD, rendering the pug at 448x448, and we see that it still looks identical to the "stock" SD (and still about 5% faster).
![pug4](https://user-images.githubusercontent.com/28718864/189553116-d7523a83-0c03-4ec9-85e8-e1e8397c77bc.png)

And Image Pug5, is rendered only on Doggettx. I can not render a 576x576 image with "stock" SD, however with Doggettx, I can render a 768x768, as seen by Pug5. And in terms of render time, it was around 1.5it/s, so while slower, it was not a lot slower.
![pug5](https://user-images.githubusercontent.com/28718864/189553182-34c33db2-4cfd-440a-9396-2dc8863a9738.png)

Another advantage of Doggettx that I can't take full advantage of on my 2060 with 6gb of vRAM, but a card with 8 or 12 gigs of ram can. Is that you can run multiple images at the same time (batch size), increasing your render speed by a relevant margin.

So @thedyze is correct, that changing the resolution effects the image output. This is not because of doggettx doing anything, but simply changing the resolution effects the way the AI generates the noise/seed. A seed can only be considered consistent between renders of the same resolution.

So Yes, Doggettx, looks different at higher resolutions than "stock", however even with stock if you were to raise or lower the resolutions on stock, within your memory capability, you would get the exact same issue. This is not an issue of Doggettx, as much as it is a flaw in how SD itself imagines the images.

TL;DR
I got the exact same image at 448x448 and 512x512, with both Stock and Doggettx. 512x512 is all my video card can do with stock, however Doggettx, is capable of rendering 768x768.
Doggettx, also appears to be about 5% faster.

EDIT: If I add [THIS](https://github.com/Doggettx/stable-diffusion/commit/a1fbe55f85dd6e7e4fdb3c9081f8f272e3233b59#diff-8909fc7ab73828e8cf431ad0448fd14a2c22b8a5a71f2bfffcc7b7b6ad7530b9) change to the webui.py, I am able to get even better memory usage. (I can render 2 images at 512x512, using this tweak. While still getting the same speed when rendering a single 512x512) (meaning the "tweak" only slows down stuff, that would be impossible on "stock".
Attached image is pug6, this is running BOTH of Doggettx tweaks, including the one mentioned in the edit as well.
![pug6](https://user-images.githubusercontent.com/28718864/189557140-346ee1b5-fa77-40c7-b48a-537d792894d5.png)

**Reply by kwaad2** [reply]
> Ok Doggettx gave me a hint that higher guidance was needed at higher resolution. Been experimenting with 40-50 now, and am getting better results. I've also been using low res low guidance txt2img images as input in img2img. It seems this does not work that well at higher resolutions, a more detailed input image creates a lot better results.
> 
> So may not be a problem with his files after all.

I never tried higher guidance! I'll have to give this a try also!

edit: Also, I spent forever writing the the earlier response. lol.

**Reply by thedyze** [reply]
Thanks for taking the time :). Got my suspicions confirmed.

I had to compensate with lower de-noise values when going really high guidance. 
Hopefully results at higher resolutions will improve with the 1.5 model.

**Comment by twothe** [comment]
What is the benefit of the optimized version?

**Reply by kwaad2** [reply]
I would be for integrating this update. It essentially makes the optimized versions irrelevant. 

If I render 1, 512x512 image at a time, it uses all of the ram that my RTX 2060 has. (And it runs fast)
I can even render a 768x768. Which would require me to have the memory optimizations turned on. Which slows down all of my rendering. And amusingly, I still get comparable performance when rendering at 768 x 768.

The only real issue, and I'm not sure how the programming pipeline works, I was actually able to render a 1024 x 1024, 8 sample. (So 8 1024x1024 images at once) but oddly, as stable diffusion is finishing the render/iterations, but it runs out of memory when finishing saving the results! However this mod is capable of processing iterations (partially) on a 6gb VRAM system that even 80gb of VRAM can't touch (without the mod)! I would guess that it has something to do with the way the images are being assembled/output, which might be a relatively easy fix?

Anyway, my point is, it's really fast rendering a single low res image, and it uses the perfect amount of memory as you scale up. Thus allowing the system to work on both low RAM and high RAM systems optimally, without needing tweaks.

**Reply by mchaker** [reply]
I did some tests of the Doggettx optimizations here: https://github.com/sd-webui/stable-diffusion-webui/issues/673#issuecomment-1241075800
